src/widget/mod.rs

```diff
@@ -714,44 +746,63 @@ fn set_widget<'a, C, W>(widget: W, idx: Index, ui: &mut Ui<C>) where
         });
     }
 
-    // Update the widget's state.
-    let maybe_new_state = {
-        // Construct a UserInput for the widget.
-        let args = UpdateArgs {
-            idx: idx,
-            prev_state: &prev_state,
-            rect: rect,
-            style: &new_style,
-            ui: UiCell { ui: ui, idx: idx },
-        };
-        widget.update(args)
-    };
+    // Unwrap the widget's previous common state. If there is no previous common state, we'll
+    // use the new state in it's place.
+    let prev_common = maybe_prev_common.unwrap_or_else(|| CommonState {
+        rect: rect,
+        depth: depth,
+        drag_state: drag_state,
+        maybe_floating: maybe_floating,
+    });
 
-    // Check for whether or not the user input needs to be captured or uncaptured. We check after
-    // the widget update so that child widgets get the first oppotunity to capture input.
-    {
-        let new_state = match maybe_new_state {
-            Some(ref new_state) => new_state,
-            None => &prev_state.state
+    // Retrieve the widget's unique state and update it via `Widget::update`.
+    let (unique_state, has_state_updated) = {
+
+        // Unwrap our unique widget state. If there is no previous state to unwrap, call the
+        // `init_state` method to construct some initial state.
+        let mut unique_state = maybe_prev_unique_state.unwrap_or_else(|| widget.init_state());
+        let has_updated = {
+
+            // A wrapper around the widget's unique state in order to keep track of whether or not it
+            // has been updated during the `Widget::update` method.
+            let mut state = State {
+                state: &mut unique_state,
+                has_updated: false,
+            };
+
+            {
+                // TODO / FIXME: For some reason passing a `&mut state` directly to the `state`
+                // field of the UpdateArgs causes the borrow checker to consider state has being
+                // borrowed even past the end of this scope, making it impossible to get the
+                // `has_updated` flag out of the state. This mutable ptr cast is a temporary
+                // hack until we can work out why this happens. It is likely the problem has
+                // something to do with UpdateArgs' lifetime intersection - I attempted various
+                // other lifetime combinations but failed to get anything working that would
+                // appease the borrow checker. As far as I can see the current intersection should
+                // be fine and safe, however the borrow checker isn't convinced.
+                //
+                // - mindtree.
+                let state_mut_ptr: *mut State<W::State> = &mut state;
+                widget.update(UpdateArgs {
+                    idx: idx,
+                    state: unsafe { ::std::mem::transmute(state_mut_ptr) },
```

Comment:
This really shouldn't be required, however I couldn't find any other way to appease the borrow checker. Perhaps there is something I'm missing? Would be nice if someone could review and see if they can get it working without the unsafe. @bvssvni ?